crates/aptos-crypto/src/p256/p256_sigs.rs

```diff
@@ -0,0 +1,191 @@
+        }
+    }
+
+    /// return an all-zero signature (for test only)
+    #[cfg(any(test, feature = "fuzzing"))]
+    pub fn dummy_signature() -> Self {
+        Self::from_bytes_unchecked(&[0u8; Self::LENGTH]).unwrap()
+    }
+
+    /// Check for correct size and third-party based signature malleability issues.
+    /// This method is required to ensure that given a valid signature for some message under some
+    /// key, an attacker cannot produce another valid signature for the same message and key.
+    ///
+    /// We use the technique described in
+    /// [BIP146](https://github.com/bitcoin/bips/blob/master/bip-0146.mediawiki) to prevent
+    /// malleability of ECDSA signatures. Signatures comprise elements {R, S}, and S can be
+    /// enforced to be of canonical form by ensuring it is less than the order of the P256 curve
+    /// divided by 2. If this is not done, a value S > n/2 can be replaced by S' = n - S to form another distinct valid
+    /// signature, where n is the curve order. This check is not performed by the RustCrypto P256 library
+    /// we use
+    pub fn check_s_malleability(bytes: &[u8]) -> std::result::Result<(), CryptoMaterialError> {
+        if bytes.len() != P256_SIGNATURE_LENGTH {
+            return Err(CryptoMaterialError::WrongLengthError);
+        }
+        if !P256Signature::check_s_lt_order_half(&bytes[32..]) {
+            return Err(CryptoMaterialError::CanonicalRepresentationError);
+        }
+        Ok(())
+    }
+
+    /// Check if S < ORDER_HALF to capture invalid signatures.
+    fn check_s_lt_order_half(s: &[u8]) -> bool {
+        for i in 0..32 {
+            match s[i].cmp(&ORDER_HALF[i]) {
+                Ordering::Less => return true,
+                Ordering::Greater => return false,
+                _ => {},
+            }
+        }
+        // As this stage S == ORDER_HALF which implies a non canonical S.
+        false
+    }
+
+    /// If the signature {R,S} does not have S < n/2 where n is the Ristretto255 order, return
+    /// {R,n-S} as the canonical encoding of this signature to prevent malleability attacks. See
+    /// `check_s_malleability` for more detail 
+    pub fn make_canonical(&self) -> P256Signature {
+        if P256Signature::check_s_malleability(&self.to_bytes()[..]).is_ok() {
+            return self.clone()
+        };
+        let s = self.0.s();
+        let r = self.0.r();
+        // NonZeroScalar::try_from throws an error on being passed the curve order, so we use the
+        // order minus one, then add one later to compute (n-1) - s + 1 = n-s
+        let order_minus_one = NonZeroScalar::try_from(&ORDER_MINUS_ONE[..]).unwrap();
+        let one = NonZeroScalar::from_uint(<NistP256 as Curve>::Uint::ONE).unwrap();
+        // Dereferencing a NonZeroScalar makes it a Scalar, which implements subtraction
+        let new_s = *order_minus_one - *s + *one;
+        let new_s_nonzero = NonZeroScalar::new(new_s).unwrap();
```

Comment:
Also, curious, there isn't a `negate()` function that you can call here, instead of doing this arithmetic on your own?

Reply:
Yeah you're right. Changed the code